### PR TITLE
Dispatch messages out of the `Bus` in FIFO order

### PR DESCRIPTION
**Pull request by vlad-lubenskyi**
Addresses https://github.com/SpineEventEngine/core-java/issues/696 and https://github.com/SpineEventEngine/core-java/issues/725.

The mentioned issues have the same root cause: in a single-threaded environment the `CommandBus` doesn't acknowledge the command until dispatching is finished; thus, routing a nested command from the command handler will block command handler until all nested commands and events are handled.

This PR addresses the problem by ensuring that `Bus` will always dispatch messages in FIFO order. Every message is positively acknowledged as soon as it was posted (that is enqueued for dispatching).

### CODE REVIEW

**Comment by vlad-lubenskyi** [comment]
@armiol PTAL

File: server/src/main/java/io/spine/server/bus/Bus.java [hunk 1]
```diff
@@ -207,6 +208,16 @@ public void close() throws Exception {
         return filterChain;
     }
 
+    /**
+     * Obtains the queue of the envelopes.
+     */
+    private DispatchingQueue<E> envelopeQueue() {
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/Bus.java]: I would easily skip the `envelope` part from its name, as it's the only queue at the moment.

Also, please add some description of what's going on here. And give the `@see` reference to the `DispatchingQueue` class.

File: server/src/main/java/io/spine/server/bus/DispatchingQueue.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    private boolean dispatchingInProgress;
+
+    DispatchingQueue(DispatchAction<E> dispatchAction) {
+        this.dispatchAction = dispatchAction;
+    }
+
+    /**
+     * Enqueues the envelope for dispatching and dispatches it in its turn.
+     *
+     * <p>The given envelope is dispatched right away if and only if:
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DispatchingQueue.java]:
```
<p>Once an envelope is being added to the queue, a number of actions may be performed.
<ul>
     <li>Queue is empty and nothing is being dispatched at the moment -> the envelope is dispatched right away;</li>
     <li>Queue is non-empty and nothing is being dispatched at the moment -> the envelope is enqueued as the first item + the last item of the queue starts to be dispatched.</li>
     <li>Another envelope is being dispatched at the moment -> the given envelope is enqueued as the first item and its dispatching is postponed.</li>
</ul>

<p>Once the envelope dispatching action is finished, the next available item is picked from the queue and its dispatching is started.

...     
```

File: server/src/test/proto/spine/test/command/commands.proto [hunk 3]
```diff
@@ -43,8 +43,28 @@ message CmdStartProject {
...
+    ProjectId id = 1;
+
+    string name = 2;
+}
+
+message FCmdStartProject {
+
+    ProjectId id = 1;
+}
+
```
Comment (armiol) [inline server/src/test/proto/spine/test/command/commands.proto]: This line is redundant.

File: server/src/test/proto/spine/test/command/commands.proto [hunk 4]
```diff
@@ -43,8 +43,28 @@ message CmdStartProject {
...
+// A command to create a project to test the command posting order.
+//
+message FCmdCreateProject {
+
+    ProjectId id = 1;
+
+    string name = 2;
+}
+
+message FCmdStartProject {
```
Comment (armiol) [inline server/src/test/proto/spine/test/command/commands.proto]: If `F` stands for `First` I would say so. And its `TestEnv` factory methods starting with `f` are also unreadable. Let's make their naming more user-friendly.

And the Message description is needed.

**Comment by vlad-lubenskyi** [comment]
@armiol PTAL

File: server/src/main/java/io/spine/server/bus/DispatchingQueue.java [hunk 5]
```diff
@@ -49,14 +49,18 @@
...
+     *     <li>if nothing is being dispatched and the queue is empty the envelope is dispatched
+     *         right away;
+     *     <li>if nothing is being dispatched and the queue is non-empty the envelope is enqueued
+     *         as the first item, and the queue dispatches the last item.
      * </ul>
      *
-     * <p>Otherwise, the envelope is only enqueued and will be dispatched according to its position
-     * in the queue.
+     * <p>Once the envelope dispatching action is finished, the queue picks the next available item
+     * and starts to dispatch it, until the queue is empty.
```
Comment (armiol) [inline server/src/main/java/io/spine/server/bus/DispatchingQueue.java]: I think the comma is redundant here.

**Pull request merged**